posts/2021/01/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,264 @@
+that at least most of the iterations happen on contiguous data to take
+advantage of hierarchical cache-based memory architectures (think dimension
+coalescing and reordering for maximum data locality).
+
+A multi-dimensional tensor has a stride value for each dimension. So the stride
+that `TensorIterator` needs to use will be different depending on which
+dimension you want to iterate over. `TensorIterator` directly computes the
+strides that get passed into the loop by itself within the `build()` function.
+How exactly it computes the dimension to iterate over is something that should
+be properly understood in order to use `TensorIterator` effectively.
+
+When performing a reduction operation (see the `sum_out` code in
+[ReduceOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L433)),
+`TensorIterator` will figure out the dimensions that will be reduced depending
+on the shape of the input and output tensor, which determines how the input
+will be broadcast over the output. If you're performing a simple pointwise
+operation between two tensors (like a `addcmul` from
+[PointwiseOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/PointwiseOps.cpp#L31))
+the iteration will happen over the entire tensor, without providing a choice of
+the dimension. This allows `TensorIterator` to freely parallelize the
+computation, without guaranteeing the order of execution, since it does not
+matter anyway.
+
+For something like a cumulative sum operation, where you want be able to choose
+the dimension to reduce but iterate over multiple non-reduced dimensions
+(possibly in parallel), you must be careful to take into account two different
+strides--one for the dimension being reduced and one for all other dimensions.
+Take a look at the following example of a somewhat simplified version of the
+[cumsum
+kernel](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L47).
+
+For a 1-D input,
+[torch.cumsum](https://pytorch.org/docs/stable/generated/torch.cumsum.html?highlight=cumsum#torch.cumsum)
+calculates the sum of all elements from the beginning of the vector up to and
+including each position in the input. A 2-D input is treated as a list of
+vectors, and the cumulative sum is calculated for each vector. Higher
+dimensional inputs follow the same logic--everything is just a list of 1-D
+vectors.  So to implement a cumulative sum, we must take into account two
+different strides: the stride between elements in a vector (`result_dim_stride`
+and `self_dim_stride` in the example below) and the stride between each vector
+(`strides[0]` and `strides[1]` in the example below).
+
+```cpp
+// A cumulative sum's output is the same size as the input
+at::Tensor result = at::empty_like(self);
+
+at::TensorIteratorConfig iter_config;
+auto iter = iter_config
+  .check_all_same_dtype(false)
+  .resize_outputs(false)
```

Comment:
Not needed here, probably. But if you restride your tensors with zero stride, you will need to use `.set_check_mem_overlap(false)`, because Tensors with zero strides are considered having overlaps in TensorIterator.

Reply:
`resize_outputs` is required when using `declare_static_shape`. I get this error if I don't include it:

```
terminate called after throwing an instance of 'c10::Error'
  what():  resize_outputs() must be called before declare_static_shape(...)
Exception raised from declare_static_shape at /pytorch/aten/src/ATen/TensorIterator.cpp:100
```

Reply:
The comment was about the memory overlap check, sorry for the confusion.